src/coap_gnutls.c

```diff
@@ -1904,6 +1919,16 @@ coap_dtls_receive(coap_session_t *c_session,
       coap_session_connected(c_session);
     }
     else {
+      if (ssl_data->pdu_len) {
+        /* Do the handshake again incase of internal timeout */
+        ret = do_gnutls_handshake(c_session, g_env);
+        if (ret == 1) {
+          /* Just connected, so send the data */
+           coap_session_connected(c_session);
+        } else {
+          ret = -1;
+        }
+      }
       ret = -1;
```

Comment:
Good catch.  Actually the last line needs to be wrapped in an `else` in case the `if (ssl_data->pdu_len)` is false.  I have moved the last line up to just before the  `if (ssl_data->pdu_len)` and removed the now redundant else statement.  The code now reads
````
    else {
      ret = -1;
      if (ssl_data->pdu_len) {
        /* Do the handshake again incase of internal timeout */
        ret = do_gnutls_handshake(c_session, g_env);
        if (ret == 1) {
          /* Just connected, so send the data */
           coap_session_connected(c_session);
        }
      }
    }
````
Code change pushed.